pkg/auth/providers/azure/clients/ms_graph_client.go

```diff
@@ -31,197 +36,308 @@ const (
-func (c azureMSGraphClient) ListGroups(filter string) ([]v3.Principal, error) {
-	groups, _, err := c.groupClient.List(context.Background(), odata.Query{Filter: filter})
+func (c AzureMSGraphClient) ListGroups(filter string) ([]v3.Principal, error) {
+	logrus.Debugf("[%s] ListGroups %s", providerLogPrefix, filter)
+	result, err := c.GraphClient.Groups().Get(context.Background(), &msgraphgroups.GroupsRequestBuilderGetRequestConfiguration{
+		QueryParameters: &msgraphgroups.GroupsRequestBuilderGetQueryParameters{
+			Filter: &filter,
+		}})
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("listing groups: %w", err)
 	}
-	var principals []v3.Principal
-	for _, u := range *groups {
-		principal, err := c.groupToPrincipal(u)
-		if err != nil {
-			return nil, err
-		}
-		principals = append(principals, principal)
+
+	pageIterator, err := msgraphcore.NewPageIterator[models.Groupable](
+		result, c.GraphClient.GetAdapter(), models.CreateGroupCollectionResponseFromDiscriminatorValue)
+	if err != nil {
+		return nil, fmt.Errorf("iterating over group list: %w", err)
 	}
-	return principals, err
+
+	var groups []v3.Principal
+	err = pageIterator.Iterate(context.Background(), func(group models.Groupable) bool {
+		groups = append(groups, groupToPrincipal(group))
+		return true
+	})
+
+	return groups, err
+
 }
 
 // ListGroupMemberships takes a user ID and fetches the user's group principals as string IDs from the Microsoft Graph API.
-func (c azureMSGraphClient) ListGroupMemberships(id string) ([]string, error) {
-	groups, _, err := c.userClient.ListGroupMemberships(context.Background(), id, odata.Query{})
+func (c AzureMSGraphClient) ListGroupMemberships(userID string) ([]string, error) {
+	logrus.Debugf("[%s] ListGroupMemberships %s", providerLogPrefix, userID)
+	var groupIDs []string
+	err := c.listGroupMemberships(userID, func(g *models.Group) {
+		if id := g.GetId(); id != nil {
+			groupIDs = append(groupIDs, *id)
+		}
+	})
 	if err != nil {
 		return nil, err
 	}
-	var names []string
-	for _, g := range *groups {
-		if g.ID != nil && g.DisplayName != nil && g.SecurityEnabled != nil {
-			names = append(names, *g.ID)
-		}
+
+	return groupIDs, nil
+}
+
+func (c AzureMSGraphClient) listGroupMemberships(userID string, f func(*models.Group)) error {
+	result, err := c.GraphClient.Users().
+		ByUserId(userID).
+		MemberOf().
+		Get(context.Background(), nil)
+	if err != nil {
+		return fmt.Errorf("listing group memberships: %w", err)
 	}
-	return names, nil
+
+	pageIterator, err := msgraphcore.NewPageIterator[models.DirectoryObjectable](
+		result, c.GraphClient.GetAdapter(),
+		models.CreateDirectoryObjectCollectionResponseFromDiscriminatorValue)
+	if err != nil {
+		return fmt.Errorf("iterating over group membership list: %w", err)
+	}
+
+	err = pageIterator.Iterate(context.Background(), func(do models.DirectoryObjectable) bool {
+		group, ok := do.(*models.Group)
+		if !ok {
+			return true
+		}
+		f(group)
+
+		return true
+	})
```

Comment:
I have the feeling that we should add the `context.Context` to the exported `ListGroupMemberships` func, or at least reuse the same context used by the previous "Get User" in the iterator. This to avoid the issue with very long iterations. What do you think?

Also we could probably log the real type if this fails, since we are just running through (it should not happen though)

```go
group, ok := do.(*models.Group)
if !ok {
	// log wrong type assertion
	return true
}
```

Reply:
Yeah, I definitely think that it should use a passed in context.Context, it's almost always a smell if you're creating contexts inside functions that are being called.